tests/suites/test_suite_bignum_mod.function

```diff
@@ -80,7 +80,178 @@ exit:
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BAD_INPUT_DATA,
+                mbedtls_mpi_mod_write( &r, &m2, r_buff, 1, endian ) );
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BAD_INPUT_DATA,
+                mbedtls_mpi_mod_write( &rn, &m, r_buff, 1, endian ) );
+
+    /* Fail for r_limbs < m->limbs */
+    r.limbs = m.limbs - 1;
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BAD_INPUT_DATA,
+                mbedtls_mpi_mod_read( &r, &m, r_buff, 1, endian ) );
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BAD_INPUT_DATA,
+                mbedtls_mpi_mod_write( &rn, &m, r_buff, 1, endian ) );
+    r.limbs = r_limbs;
+
+    /* Fail if input_r >= modulo m */
+    /* input_r = modulo */
+    memset( r_buff, 0xfe, buff_bytes );
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BAD_INPUT_DATA,
+                mbedtls_mpi_mod_read( &r, &m, r_buff, 1, endian ) );
+
+    /* input_r > modulo */
+    memset( r_buff, 0xff, buff_bytes );
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BAD_INPUT_DATA,
+                mbedtls_mpi_mod_read( &r, &m, r_buff, 1, endian ) );
+
+    /* Data too large to fit */
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BUFFER_TOO_SMALL,
+                mbedtls_mpi_mod_read( &r, &m, r_buff, buff_bytes, endian ) );
+
+    /* Read the two limbs input data into a larger modulus and residue */
+    TEST_EQUAL( 0, mbedtls_mpi_mod_modulus_setup( &m2, N2, n2_limbs,
+                MBEDTLS_MPI_MOD_REP_MONTGOMERY ) );
+    rn.p = R2;
+    rn.limbs = r2_limbs;
+    TEST_EQUAL( MBEDTLS_ERR_MPI_BUFFER_TOO_SMALL,
+                mbedtls_mpi_mod_write( &rn, &m2, r_buff, 1, endian ) );
+
+exit:
+    mbedtls_mpi_mod_modulus_free( &m );
+    mbedtls_mpi_mod_modulus_free( &m2 );
+    mbedtls_free( N );
+    mbedtls_free( R );
+    mbedtls_free( N2 );
+    mbedtls_free( R2 );
+    mbedtls_free( r_buff );
+}
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mpi_mod_io( char * input_N, data_t * input_A, int endian )
+{
+    mbedtls_mpi_uint *N = NULL;
+    mbedtls_mpi_uint *R = NULL;
+    unsigned char *r_buff = NULL;
+    mbedtls_mpi_mod_modulus m;
+    mbedtls_mpi_mod_residue r;
+    size_t n_limbs, n_bytes, a_bytes;
+
+    /* Read inputs */
+    TEST_EQUAL( 0, mbedtls_test_read_mpi_core( &N, &n_limbs, input_N ) );
+    n_bytes = n_limbs * sizeof( mbedtls_mpi_uint );
+    a_bytes = input_A->len * sizeof( char );
+
+    /* Allocate the memory for intermediate data structures */
+    ASSERT_ALLOC( R, n_bytes );
+    ASSERT_ALLOC( r_buff, a_bytes );
+
+    /* Test that input's size is not greater to modulo's */
+    TEST_LE_U(a_bytes, n_bytes );
+
+    /* Init Structures */
+    mbedtls_mpi_mod_modulus_init( &m );
+    TEST_EQUAL( 0, mbedtls_mpi_mod_modulus_setup( &m, N, n_limbs,
+                MBEDTLS_MPI_MOD_REP_MONTGOMERY ) );
+
+    /* Enforcing p_limbs >= m->limbs */
+    TEST_EQUAL( 0, mbedtls_mpi_mod_residue_setup( &r, &m, R, n_limbs ) );
+
+    TEST_EQUAL( 0, mbedtls_mpi_mod_read( &r, &m, input_A->x, input_A->len,
+                                         endian ) );
+
```

Comment:
Missing: testing write into larger/smaller buffer. To be done with the same data, enumerate several different buffer sizes to write to, remember to allocate a tight size for the output buffer each time. For smaller buffers, the expected outcome depends on the number of leading zeros, so the test code will need to count (or get it from the test data but I think that's just more difficult if you take the writing and review time into account).